src/Data/AffineSpace.hs

```diff
@@ -122,3 +140,58 @@ instance (AffineSpace p) => AffineSpace (a -> p) where
   type Diff (a -> p) = a -> Diff p
   (.-.)              = liftA2 (.-.)
   (.+^)              = liftA2 (.+^)
+
+
+
+newtype GenericDiff p = GenericDiff (Diff (Gnrx.Rep p ()))
+       deriving (Generic)
+
+instance AdditiveGroup (Diff (Gnrx.Rep p ())) => AdditiveGroup (GenericDiff p)
+instance VectorSpace (Diff (Gnrx.Rep p ())) => VectorSpace (GenericDiff p)
+instance InnerSpace (Diff (Gnrx.Rep p ())) => InnerSpace (GenericDiff p)
+instance HasBasis (Diff (Gnrx.Rep p ())) => HasBasis (GenericDiff p)
+
+data AffineDiffProductSpace f g p = AffineDiffProductSpace
```

Comment:
The issue with re-combining `Diff`s via `:*:` is again that that is _lifted_. So, unless I'm missing some trick, we'd need a `Diff' :: (* -> *) -> * -> *` transformer and then have
```haskell
  type Diff ((f:*:g) p) = (Diff' f :*: Diff' g) p
```
I suspect this would require the route of having an entire lifted `Affine'` class, and as I said, I don't quite see the point.

As to strictness – yeah, I did wonder whether this is the right thing.
<br>Ideally, we'd of course adapt the same strictness properties of the original affine space type, however I don't see how that could be done. The strictness is encoded not in `:*:` but in the `S1` meta layer, and even if we wrapped the `Diff` back in such a layer I'm not sure whether it would accomplish the goal. Do these layers even affect semantics of the type at all? I don't think so; they just descibe the properties of an actually existing `data` structure. These constructors are meaningless if they're not the `Rep` of such an actual `data`, but if we somehow re-wrap the `Diff`s then the result will _not_ be the `Rep` of any type that we have control over.

Making it strict (or even `{-# UNBOX #-}`?) seems a good decision for all the performance-hungry applications in which vector spaces are so interesting. I certainly wouldn't feel good about it if strict affine spaces would end up with lazy `Diff`s.

However, I reckon it would be possible to also account for lazy fields. But it would require yet another set of `newtype`s to use in the `AffineSpace (f p) => AffineSpace (Gnrx.M1 i c f p)` instance. Is that worth it? I'm not sure.